File: csi_camera_ros2/csi_camera_ros2/image_viewer_node.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
import cv2
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError

class ImageViewerNode(Node):
    def __init__(self):
        super().__init__('image_viewer_node')
        self.bridge = CvBridge()
        # Subscribe to the raw image topic published by csi_camera_node
        # Using default reliable QoS for simplicity first
        self.subscription = self.create_subscription(
            Image,
            '/csi_camera_0/image_raw', # Assuming default sensor_id=0
            self.image_callback,
            10) # Default QoS depth
        self.subscription  # prevent unused variable warning
        self.get_logger().info("Image viewer node started, subscribing to /csi_camera_0/image_raw")
        self.window_title = "ROS Image Viewer"
        cv2.namedWindow(self.window_title, cv2.WINDOW_AUTOSIZE)

    def image_callback(self, msg):
        # self.get_logger().debug('Received image') # Uncomment for verbose debugging
        try:
            # Convert ROS Image message to OpenCV image
            cv_image = self.bridge.imgmsg_to_cv2(msg, "bgr8")
        except CvBridgeError as e:
            self.get_logger().error(f'CV Bridge Error: {e}')
            return

        # Display image
        cv2.imshow(self.window_title, cv_image)
        key = cv2.waitKey(1) & 0xFF # Need waitKey for imshow to work

        # Check if window was closed (using AUTOSIZE property as a proxy)
        # Or if 'q' or ESC is pressed
        if cv2.getWindowProperty(self.window_title, cv2.WND_PROP_AUTOSIZE) < 0 or \
           key == ord('q') or key == 27:
            self.get_logger().info("Closing image viewer window.")
            cv2.destroyWindow(self.window_title)
            # Optional: shutdown the node when window is closed
            # rclpy.shutdown()
            # Or just destroy the subscription/node components if you want the node to exit cleanly
            self.destroy_subscription(self.subscription)
            # Need a way to signal the main spin loop to exit, or handle node destruction better
            # For simplicity now, we just stop displaying. A real app might shut down.


def main(args=None):
    rclpy.init(args=args)
    image_viewer_node = ImageViewerNode()
    try:
        # Need spin to keep node alive and process callbacks
        rclpy.spin(image_viewer_node)
    except KeyboardInterrupt:
        pass
    finally:
        # Cleanup OpenCV window just in case
        cv2.destroyAllWindows()
        # Destroy the node explicitly
        image_viewer_node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == '__main__':
    main()

File: csi_camera_cpp/src/image_viewer_node.cpp
#include "csi_camera_cpp/image_viewer_node.hpp"
#include <rclcpp_components/register_node_macro.hpp>
#include <opencv2/highgui.hpp> // For cv::imshow, cv::waitKey, etc.
#include <cv_bridge/cv_bridge.h> // For image conversion

namespace csi_camera_cpp
{

ImageViewerNode::ImageViewerNode(const rclcpp::NodeOptions & options)
: Node("image_viewer_node", options)
{
    RCLCPP_INFO(this->get_logger(), "Initializing ImageViewerNode...");

    // Create subscription using image_transport
    // Use default QoS for now, can customize later if needed
    // Using "raw" transport explicitly
    sub_ = image_transport::create_subscription(
        this, "image_raw", // Topic name relative to node namespace/name
        std::bind(&ImageViewerNode::image_callback, this, std::placeholders::_1),
        "raw",
        rmw_qos_profile_sensor_data); // Use sensor data QoS profile for potentially lower latency

    RCLCPP_INFO(this->get_logger(), "Subscribed to %s", sub_.getTopic().c_str());
    cv::namedWindow(window_name_, cv::WINDOW_AUTOSIZE);
    RCLCPP_INFO(this->get_logger(), "ImageViewerNode initialized.");
}

ImageViewerNode::~ImageViewerNode()
{
    RCLCPP_INFO(this->get_logger(), "Shutting down ImageViewerNode...");
    cv::destroyWindow(window_name_);
    RCLCPP_INFO(this->get_logger(), "ImageViewerNode shutdown complete.");
}

void ImageViewerNode::image_callback(const sensor_msgs::msg::Image::ConstSharedPtr & msg)
{
    try {
        // Convert ROS Image message to OpenCV image.
        // Use toCvShare to avoid copying if possible (though display might force copy)
        cv_bridge::CvImageConstPtr cv_ptr = cv_bridge::toCvShare(msg, sensor_msgs::image_encodings::BGR8);

        // Display image
        cv::imshow(window_name_, cv_ptr->image);

    } catch (const cv_bridge::Exception& e) {
        RCLCPP_ERROR(this->get_logger(), "CV Bridge error: %s", e.what());
        return; // Return on error before waitKey
    } catch (const cv::Exception& e) {
        RCLCPP_ERROR(this->get_logger(), "OpenCV display/processing error: %s", e.what());
        return; // Return on error before waitKey
    }

    // Ensure waitKey is called *after* imshow and outside the main try block
    int key = cv::waitKey(1); // Use waitKey(1) for minimal delay

    // Check if window was closed or 'q'/ESC pressed
    // Use getWindowProperty check first as waitKey might return -1 if no key is pressed
    if (cv::getWindowProperty(window_name_, cv::WND_PROP_VISIBLE) < 1 || (key != -1 && (key == 'q' || key == 27))) {
         RCLCPP_INFO(this->get_logger(), "Close request received, shutting down viewer.");
         cv::destroyWindow(window_name_);
         // Signal main thread to exit or shutdown node cleanly
         // For simplicity, we rely on Ctrl+C in the launch terminal for now
         // Or could call rclcpp::shutdown();
         // Or potentially just stop the timer/subscription if the node should continue running
    }
    // Removed duplicated catch blocks
}

} // namespace csi_camera_cpp

// Register the component with class_loader
RCLCPP_COMPONENTS_REGISTER_NODE(csi_camera_cpp::ImageViewerNode)
